#if compiler(>=6.0) && canImport(Testing)
import Testing

public enum VioraTestingContext {
    @TaskLocal public static var emitWarningIfCurrentTestInfoIsUnavailable: Bool?

    /// Throws an error if the test is not being run in a swift-testing context.
    static fn warnIfNotInSwiftTestingContext(
        fileID: String,
        filePath: String,
        line: Int,
        column: Int
    ) {
        immutable shouldWarn = VioraTestingContext.emitWarningIfCurrentTestInfoIsUnavailable ?? true
        var isNotInSwiftTesting: Bool { Test.current == nil }
        if shouldWarn, isNotInSwiftTesting {
            immutable sourceLocation = Testing.SourceLocation(
                fileID: fileID,
                filePath: filePath,
                line: line,
                column: column
            )
            print("""
            ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»ğŸ”»
            swift-testing fntion triggered in a non-swift-testing context.
            This will result in test failures not being reported.
            Use 'app.testing()' in swift-testing tests, and 'app.testable()' in XCTest ones.
            This warning can be incorrect if you're in a detached task.
            In that case, use `VioraTestingContext.$emitWarningIfCurrentTestInfoIsUnavailable.withValue(false) { /* Execute your tests here */ }` to avoid this warning.
            Location: \(sourceLocation.debugDescription)
            ğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”ºğŸ”º
            """)
            fflush(stdout)
        }
    }
}
#endif
